Frameworks/CoreData/NSManagedObjectModel-XMLParsing.mm

```diff
@@ -0,0 +1,282 @@
+static NSString* const _kNSManagedObjectModelEntityTag = @"entity";
+static NSString* const _kNSManagedObjectModelEntityAttributeTag = @"attribute";
+static NSString* const _kNSManagedObjectModelEntityRelationshipTag = @"relationship";
+static NSString* const _kNSManagedObjectModelUserInfoTag = @"userInfo";
+static NSString* const _kNSManagedObjectModelUserInfoEntryTag = @"entry";
+static NSString* const _kNSManagedObjectModelEntityCompoundIndexTag = @"compoundIndex";
+static NSString* const _kNSManagedObjectModelEntityCompoundIndexesTag = @"compoundIndexes";
+static NSString* const _kNSManagedObjectModelEntityCompoundIndexIndexTag = @"index";
+static NSString* const _kNSManagedObjectModelEntityUniquenessConstraintTag = @"uniquenessConstraint";
+static NSString* const _kNSManagedObjectModelEntityUniquenessConstraintsTag = @"uniquenessConstraints";
+static NSString* const _kNSManagedObjectModelEntityUniquenessConstraintConstraintTag = @"constraint";
+
+// To save ourselves some string compares:
+static NSString* _uniqueTag(NSString* tag) {
+    static StrongId<NSSet> _tagSet = [NSSet setWithObjects:_kNSManagedObjectModelTag,
+                                                           _kNSManagedObjectModelEntityTag,
+                                                           _kNSManagedObjectModelEntityAttributeTag,
+                                                           _kNSManagedObjectModelEntityRelationshipTag,
+                                                           _kNSManagedObjectModelUserInfoTag,
+                                                           _kNSManagedObjectModelUserInfoEntryTag,
+                                                           _kNSManagedObjectModelEntityCompoundIndexTag,
+                                                           _kNSManagedObjectModelEntityCompoundIndexesTag,
+                                                           _kNSManagedObjectModelEntityUniquenessConstraintTag,
+                                                           _kNSManagedObjectModelEntityUniquenessConstraintsTag,
+                                                           nullptr];
+    return [_tagSet member:tag];
+}
+
+static Class _classForElementName(NSString* elementName) {
+    static StrongId<NSDictionary<NSString*, Class<_NSCDXMLCoding>>> _mappings = @{
+        _kNSManagedObjectModelTag : [NSManagedObjectModel class],
+        _kNSManagedObjectModelEntityTag : [NSEntityDescription class],
+        _kNSManagedObjectModelEntityAttributeTag : [NSAttributeDescription class],
+        _kNSManagedObjectModelEntityRelationshipTag : [NSRelationshipDescription class],
+    };
+    return [_mappings objectForKey:elementName];
+}
+
+// XML data model parsing is done via recursive descent and state management. Each element
+// we encounter can push another state onto the stack. The state dictates what state transitions
+// are available and what is to be done when the state is exited.
```

Comment:
need a warning here that the states have to be in specific order for the state machine to work.  In case someone gets adventurous later on to change this code :).